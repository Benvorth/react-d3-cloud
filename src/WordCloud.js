import { Component, PropTypes } from 'react';
import ReactFauxDom from 'react-faux-dom';
import * as d3 from 'd3';
import cloud from 'd3-cloud';

const fill = d3.scaleOrdinal(d3.schemeCategory10);

class WordCloud extends Component {
  static defaultProps = {
    width: 700,
    height: 600,
    font: 'serif',
  }

  componentWillMount() {
    this.wordCloud = ReactFauxDom.createElement('div');
  }

  render() {
    const { data, width, height, font } = this.props;
    const wordCounts = data.map(
      text => ({ ...text })
    );
    const defaultFontSizeMapper = (word) => {
      const max = 100;
      const min = 10;
      const wordCountSize = wordCounts.map(w => w.value);
      const minSize = Math.min(...wordCountSize);
      const maxSize = Math.max(...wordCountSize);
      return min + (max - min) * (word.value - minSize) / (maxSize - minSize);
    };
    const layout = cloud()
      .size([width, height])
      .font(font)
      .words(wordCounts)
      .padding(5)
      .rotate(() => 0)
      .fontSize(defaultFontSizeMapper)
      .on('end', words => {
        d3.select(this.wordCloud)
          .append('svg')
          .attr('width', layout.size()[0])
          .attr('height', layout.size()[1])
          .append('g')
          .attr('transform', `translate(${layout.size()[0] / 2},${layout.size()[1] / 2})`)
          .selectAll('text')
          .data(words)
          .enter()
          .append('text')
          .style('font-size', d => `${d.size}px`)
          .style('font-family', 'Impact')
          .style('fill', (d, i) => fill(i))
          .attr('text-anchor', 'middle')
          .attr('transform',
            d => `translate(${[d.x, d.y]})rotate(${d.rotate})`
          )
          .text(d => d.text);
      });

    layout.start();

    return this.wordCloud.toReact();
  }
}

WordCloud.propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({
    text: PropTypes.string.isRequired,
    value: PropTypes.number.isRequired,
  })).isRequired,
  width: PropTypes.number,
  height: PropTypes.number,
  font: PropTypes.string,
};

export default WordCloud;
