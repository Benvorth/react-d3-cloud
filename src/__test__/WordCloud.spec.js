
import React from 'react';
import renderer from 'react-test-renderer';

import WordCloud from '../WordCloud';

describe('index.js', () => {
  const data = [
    { text: 'Hey', value: 1000 },
    { text: 'lol', value: 200 },
    { text: 'cool', value: 1 },
    { text: 'duck', value: 10 },
  ];

  it('should contain all words', () => {
    const originalRandom = Math.random;
    Math.random = jest.fn().mockReturnValue(0.5);
    const component = renderer.create(
      <WordCloud data={data} />
    ).toJSON();
    expect(component).toMatchSnapshot();
    Math.random = originalRandom;
  });
});
