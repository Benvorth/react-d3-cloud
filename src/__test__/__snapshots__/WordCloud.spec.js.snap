exports[`index.js should contain all words 1`] = `
<div
  style={Object {}}>
  <svg
    height={600}
    style={Object {}}
    width={700}>
    <g
      style={Object {}}
      transform="translate(350,300)">
      <text
        style={
          Object {
            "fill": "#1f77b4",
            "fontFamily": "Impact",
            "fontSize": "100px",
          }
        }
        textAnchor="middle"
        transform="translate(0,0)rotate(0)">
        Hey
      </text>
      <text
        style={
          Object {
            "fill": "#ff7f0e",
            "fontFamily": "Impact",
            "fontSize": "27px",
          }
        }
        textAnchor="middle"
        transform="translate(-7,32)rotate(0)">
        lol
      </text>
      <text
        style={
          Object {
            "fill": "#2ca02c",
            "fontFamily": "Impact",
            "fontSize": "10px",
          }
        }
        textAnchor="middle"
        transform="translate(-47,18)rotate(0)">
        cool
      </text>
      <text
        style={
          Object {
            "fill": "#d62728",
            "fontFamily": "Impact",
            "fontSize": "10px",
          }
        }
        textAnchor="middle"
        transform="translate(-19,48)rotate(0)">
        duck
      </text>
    </g>
  </svg>
</div>
`;
